src/category_theory/sites/sieves.lean

```diff
@@ -0,0 +1,250 @@
+  generate S ≤ S' ↔ S ≤ S'.get_arrow_set :=
+⟨λ H g hg,
+  begin
+    have : over.mk g.hom = g,
+      cases g, dsimp [over.mk],
+      congr' 1, apply subsingleton.elim,
+    rw ← this at *,
+    exact H _ _ (generate_sets.basic hg),
+  end,
+λ ss Y f hf,
+begin
+  induction hf,
+  case basic : X g hg { exact ss hg },
+  case close : Y Z f g hf₁ hf₂ { exact S'.downward_closed hf₂ _ },
+end⟩
+
+/-- Show that there is a galois insertion (generate, get_arrow_set). -/
+def gi_generate : galois_insertion (generate : set (over X) → sieve X) get_arrow_set :=
+{ gc := sets_iff_generate,
+  choice := λ 𝒢 _, generate 𝒢,
+  choice_eq := λ _ _, rfl,
+  le_l_u := λ S Y f hf, generate_sets.basic hf }
+
+/-- Given a morphism `h : Y ⟶ X`, send a sieve S on X to a sieve on Y
+    as the inverse image of S with `_ ≫ h`.
+    That is, `sieve.pullback S h := (≫ h) '⁻¹ S`. -/
+def pullback (S : sieve X) (h : Y ⟶ X) : sieve Y :=
+{ arrows := λ Y sl, S.arrows (sl ≫ h),
+  downward_closed := λ Z W f g h, by simp [g] }
+
+@[simp] lemma mem_pullback (h : Y ⟶ X) {f : Z ⟶ Y} :
+  (pullback S h).arrows f ↔ S.arrows (f ≫ h) := iff.rfl
+
+/--
+Push a sieve `R` on `Y` forward along an arrow `f : Y ⟶ X`: `gf : Z ⟶ X`
+is in the sieve if `gf` factors through some `g : Z ⟶ Y` which is in `R`.
+-/
+def comp (R : sieve Y) (f : Y ⟶ X) : sieve X :=
+{ arrows := λ Z gf, ∃ g, g ≫ f = gf ∧ R.arrows g,
+  downward_closed := λ Z₁ Z₂ g ⟨j, k, z⟩ h, ⟨h ≫ j, by simp [k], by simp [z]⟩ }
+
+@[simp]
+lemma mem_comp_of_comp {R : sieve Y} {Z : C} {g : Z ⟶ Y} (hg : R.arrows g) (f : Y ⟶ X) :
+  (R.comp f).arrows (g ≫ f) :=
+⟨g, rfl, hg⟩
+
+/-- Pullback is monotonic -/
+lemma pullback_le_map {S R : sieve X} (Hss : S ≤ R) (f : Y ⟶ X) : pullback S f ≤ pullback R f :=
+λ Z H, Hss _ _
+
+lemma pullback_top {f : Y ⟶ X} : pullback ⊤ f = ⊤ :=
+top_unique (λ _ g, id)
+
+lemma pullback_comp {f : Y ⟶ X} {g : Z ⟶ Y} (S : sieve X) : S.pullback (g ≫ f) = (S.pullback f).pullback g :=
+by simp [sieve.ext_iff]
+
+lemma pullback_inter {f : Y ⟶ X} (S R : sieve X) : (S ⊓ R).pullback f = S.pullback f ⊓ R.pullback f :=
+by simp [sieve.ext_iff]
+
+lemma le_pullback_comp {R : sieve Y} {f : Y ⟶ X} :
+  R ≤ pullback (comp R f) f :=
+λ Z g h, ⟨g, rfl, h⟩
+
+/-- If the identity arrow is in a sieve, the sieve is maximal. -/
+lemma id_mem_iff_eq_top : S.arrows (𝟙 X) ↔ S = ⊤ :=
+⟨λ h, top_unique $ λ Y f _, by simpa using downward_closed _ h f,
+ λ h, h.symm ▸ trivial⟩
+
+lemma pullback_eq_top_iff_mem (f : Y ⟶ X) : S.arrows f ↔ S.pullback f = ⊤ :=
+by rw [← id_mem_iff_eq_top, mem_pullback, category.id_comp]
+
+/-- A sieve induces a presheaf. -/
+@[simps]
+def functor (S : sieve X) : Cᵒᵖ ⥤ Type v :=
+{ obj := λ Y, {g : Y.unop ⟶ X // S.arrows g},
+  map := λ Y Z f g, ⟨f.unop ≫ g.1, downward_closed _ g.2 _⟩ }
+
+/--
+If a sieve S is contained in a sieve T, then we have a morphism of presheaves on their induced
+presheaves.
+-/
+@[simps]
+def le_functor {S T : sieve X} (h : S ≤ T) : S.functor ⟶ T.functor :=
```

Comment:
```suggestion
def functor_hom_of_le {S T : sieve X} (h : S ≤ T) : S.functor ⟶ T.functor :=
```
A bit longer, but maybe clearer?

Reply:
Good point - how about `nat_trans_of_le`? Writing `functor_hom` feels a little silly to me when we already have `nat_trans`